### PR TITLE
Optional proxy of Prometheus from outside the cluster

### PR DESCRIPTION
**Pull request by yaacov**
**Description**:
It will be nice to have an option to proxy prometheus from outside the cluster for devel.

This PR adds the option to proxy Prometheus and AlertManager when running outside the
cluster.

It adds new CLI options:
```
./bin/bridge --help
...
  -k8s-mode-off-cluster-alertmanager string
    	DEV ONLY. URL of the cluster's AlertManager server.
  -k8s-mode-off-cluster-prometheus string
    	DEV ONLY. URL of the cluster's Pormetheus server.

...
```

**Usage**:
Already integrated when running dev scripts:
```
source ./contrib/oc-environment.sh
./bin/bridge
```
And
```
./examples/run-bridge.sh
```

Ref:
https://docs.google.com/document/d/1epyQuj0AqL77zCLuea6T6WUKq-v47NbT9v24pOBFS1A/edit

https://jira.coreos.com/browse/CONSOLE-737

### CODE REVIEW

**Comment by yaacov** [comment]
@dtaylor113 Hi, 
This is a way to get data from prometheus when running is dev env.
How do you run pages [1] that need to get data from prometheus on dev env ?

[1] https://github.com/openshift/console/pull/2497

**Comment by yaacov** [comment]
@rawagner @suomiy @mareklibra @vojtechszocs @spadgett Hi,

Is there a way to get data from Promethean/AlertManger in devel mode ? what am I missing something ?

File: cmd/bridge/main.go [hunk 1]
```diff
@@ -96,6 +96,9 @@ func main() {
 
 	fLoadTestFactor := fs.Int("load-test-factor", 0, "DEV ONLY. The factor used to multiply k8s API list responses for load testing purposes.")
 
+	fPrometheusDirectURL := fs.String("prometheus-direct-url", "", "DEV ONLY. Format: <http | https>://domainOrIPAddress[:port]. Example: https://prometheus.example.com.")
```
[the remark this answers is not in the document]
Comment (yaacov) [reply cmd/bridge/main.go]: :+1: Thanks !

File: README.md [hunk 2]
```diff
@@ -156,6 +156,27 @@ yarn run dev
...
+
+Monitoring features requires the bridge to run inside the cluster. For develop flow
+It is possible to simulate the proxy via direct links.
+
+#### Set a Development setup
+
+Run a web browser with web security disabled.
+```
+google-chrome --args --new-window \
+    --disable-web-security \
```
[the remark this answers is not in the document]
Comment (yaacov) [reply README.md]: :+1: removed, we now use proxy to get the data, no CORS :tada:

File: cmd/bridge/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yaacov) [reply cmd/bridge/main.go]: use the proxy now :birthday: :beers: :cake:

File: cmd/bridge/main.go [hunk 3]
```diff
@@ -68,6 +68,8 @@ func main() {
 	fK8sMode := fs.String("k8s-mode", "in-cluster", "in-cluster | off-cluster")
 	fK8sModeOffClusterEndpoint := fs.String("k8s-mode-off-cluster-endpoint", "", "URL of the Kubernetes API server.")
 	fK8sModeOffClusterSkipVerifyTLS := fs.Bool("k8s-mode-off-cluster-skip-verify-tls", false, "DEV ONLY. When true, skip verification of certs presented by k8s API server.")
+	fK8sModeOffClusterPrometheus := fs.String("k8s-mode-off-cluster-prometheus", "", "DEV ONLY. URL of the clusters Prmetheus server.")
```
Comment (kyoto) [inline cmd/bridge/main.go]: Typo: `Prmetheus` -> `Prometheus`

Comment (yaacov) [reply cmd/bridge/main.go]: Thanks !

File: cmd/bridge/main.go [hunk 4]
```diff
@@ -68,6 +68,8 @@ func main() {
 	fK8sMode := fs.String("k8s-mode", "in-cluster", "in-cluster | off-cluster")
 	fK8sModeOffClusterEndpoint := fs.String("k8s-mode-off-cluster-endpoint", "", "URL of the Kubernetes API server.")
 	fK8sModeOffClusterSkipVerifyTLS := fs.Bool("k8s-mode-off-cluster-skip-verify-tls", false, "DEV ONLY. When true, skip verification of certs presented by k8s API server.")
+	fK8sModeOffClusterPrometheus := fs.String("k8s-mode-off-cluster-prometheus", "", "DEV ONLY. URL of the clusters Prometheus server.")
```
Comment (kyoto) [inline cmd/bridge/main.go]: Typo: `clusters` -> `cluster's`

Comment (yaacov) [reply cmd/bridge/main.go]: I should wait with fixing :-)

**Comment by yaacov** [comment]
@spadgett hi, removed the need to use un secure browser, now using proxy, please re-review.

File: contrib/oc-environment.sh [hunk 5]
```diff
@@ -17,6 +17,8 @@ export BRIDGE_USER_AUTH="disabled"
 export BRIDGE_K8S_MODE="off-cluster"
 export BRIDGE_K8S_MODE_OFF_CLUSTER_ENDPOINT=$(oc whoami --show-server)
 export BRIDGE_K8S_MODE_OFF_CLUSTER_SKIP_VERIFY_TLS=true
+export BRIDGE_K8S_MODE_OFF_CLUSTER_PROMETHEUS=$(oc -n openshift-monitoring get configmap sharing-config -o json | jq .data.prometheusURL | tr -d '"')
```
Comment (spadgett) [inline contrib/oc-environment.sh]: You can avoid the `jq` requirement with
```suggestion
export BRIDGE_K8S_MODE_OFF_CLUSTER_PROMETHEUS=$(oc -n openshift-monitoring get configmap sharing-config -o jsonpath='{.data.prometheusURL}')
```

(same for below)

File: pkg/server/server.go [hunk 6]
```diff
@@ -90,18 +90,21 @@ type Server struct {
...
+	OffClusterAlertManagerURL *url.URL
 }
 
 func (s *Server) authDisabled() bool {
 	return s.Auther == nil
 }
 
 func (s *Server) prometheusProxyEnabled() bool {
-	return s.PrometheusProxyConfig != nil && s.PrometheusTenancyProxyConfig != nil
+	return (s.PrometheusProxyConfig != nil && s.PrometheusTenancyProxyConfig != nil) || s.OffClusterPrometheusURL.String() != ""
```
Comment (spadgett) [inline pkg/server/server.go]: Isn't the proxy config still set? Is this change needed?

**Comment by spadgett** [comment]
Please squash when ready to merge

**Comment by spadgett** [comment]
lgtm, just need to squash. thanks!

File: pkg/server/server.go (hunk 6 above)
Comment (yaacov) [reply pkg/server/server.go]: Yes, my bad ... fixing :+1:

**Comment by yaacov** [comment]
@spadgett do-not-merge/hold label ?
